src/charm.py

```diff
@@ -0,0 +1,352 @@
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        self._write_ue_config_file()
+
+    def _configure_sdcore_smf(self, event: EventBase) -> None:
+        """Adds pebble layer and manages Juju unit status.
+
+        Args:
+            event: Juju event
+        """
+        if not self._default_database_relation_is_created:
+            self.unit.status = BlockedStatus(
+                "Waiting for `default-database` relation to be created"
+            )
+            return
+        if not self._smf_database_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for `smf-database` relation to be created")
+            return
+        if not self._nrf_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for NRF relation to be created")
+            return
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            return
+        if not self._default_database_is_available:
+            self.unit.status = WaitingStatus(
+                "Waiting for `default-database` relation to be available"
+            )
+            return
+        if not self._smf_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for `smf-database` relation to be available")
+            return
+        if not self._nrf_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF relation to be available")
+            return
+        if not self._ue_config_file_is_written:
+            event.defer()
+            self.unit.status = WaitingStatus(
+                f"Waiting for `{UEROUTING_CONFIG_FILE}` config file to be pushed to workload container"  # noqa: W505, E501
+            )
+            return
+        if not self._config_file_is_written:
+            self._write_config_file(
+                default_database_name=DEFAULT_DATABASE_NAME,
+                default_database_url=self._smf_database_data["uris"].split(",")[0],
+                smf_database_name=SMF_DATABASE_NAME,
+                smf_url=self._smf_hostname,
+                smf_sbi_port=SMF_SBI_PORT,
+                nrf_url=self._nrf_requires.get_nrf_url(),
+                pod_ip=str(self._pod_ip),
+            )
+        self._configure_pebble()
+
+    def _configure_pebble(self) -> None:
+        """Adds layer to pebble config if the proposed config is different from the current one."""
+        self._container.add_layer("smf", self._pebble_layer, combine=True)
+        self._container.replan()
+        self.unit.status = ActiveStatus()
+
+    def _write_config_file(
+        self,
+        default_database_name: str,
+        default_database_url: str,
+        smf_database_name: str,
+        smf_url: str,
+        smf_sbi_port: int,
+        nrf_url: str,
+        pod_ip: str,
+    ) -> None:
```

Comment:
This is a lot of arguments to a single method, but I think it is necessary in this case. A good thing in such cases would be to force usage of keyword arguments, like so:
```suggestion
    def _write_config_file(
        self,
        *,
        default_database_name: str,
        default_database_url: str,
        smf_database_name: str,
        smf_url: str,
        smf_sbi_port: int,
        nrf_url: str,
        pod_ip: str,
    ) -> None:
```

Calling the method with positional arguments after that change would fail.

As this is a private method and you are already calling it with keywords arguments, this is just a small suggestion, feel free to close it.

Reply:
I like this, I've added it. I didn't know it, thanks for the comment, pretty cool!